p2p/server/server.go

```diff
@@ -235,44 +235,34 @@ func (s *Server) queueHandler(ctx context.Context, stream network.Stream) {
 
 // Request sends a binary request to the peer. Request is executed in the background, one of the callbacks
 // is guaranteed to be called on success/error.
-func (s *Server) Request(
-	ctx context.Context,
-	pid peer.ID,
-	req []byte,
-	resp func([]byte),
-	failure func(error),
-) error {
+func (s *Server) Request(ctx context.Context, pid peer.ID, req []byte) ([]byte, error) {
 	start := time.Now()
 	if len(req) > s.requestLimit {
-		return fmt.Errorf("request length (%d) is longer than limit %d", len(req), s.requestLimit)
+		return nil, fmt.Errorf("request length (%d) is longer than limit %d", len(req), s.requestLimit)
 	}
 	if s.h.Network().Connectedness(pid) != network.Connected {
-		return fmt.Errorf("%w: %s", ErrNotConnected, pid)
+		return nil, fmt.Errorf("%w: %s", ErrNotConnected, pid)
 	}
-	go func() {
-		data, err := s.request(ctx, pid, req)
-		if err != nil {
-			failure(err)
-		} else if len(data.Error) > 0 {
-			failure(errors.New(data.Error))
-		} else {
-			resp(data.Data)
-		}
-		s.logger.WithContext(ctx).With().Debug("request execution time",
-			log.String("protocol", s.protocol),
-			log.Duration("duration", time.Since(start)),
-			log.Err(err),
-		)
-		switch {
-		case s.metrics == nil:
-			return
-		case err != nil:
-			s.metrics.clientLatencyFailure.Observe(time.Since(start).Seconds())
-		case err == nil:
-			s.metrics.clientLatency.Observe(time.Since(start).Seconds())
-		}
-	}()
-	return nil
+	data, err := s.request(ctx, pid, req)
+	s.logger.WithContext(ctx).With().Debug("request execution time",
+		log.String("protocol", s.protocol),
+		log.Duration("duration", time.Since(start)),
+		log.Err(err),
+	)
+
+	took := time.Since(start).Seconds()
+	switch {
+	case s.metrics == nil:
```

Comment:
This breaks error handling when metrics are off. Fixed in the merge commit in https://github.com/spacemeshos/go-spacemesh/tree/refactor-fetch-request-upd

Reply:
It's already fixed in [0d863c3](https://github.com/spacemeshos/go-spacemesh/pull/5453/commits/0d863c30780f0cc99d1f41ee7eb775606454a0fe) :)